[foxy backport] Fix no specified namespace (#153)

jacobperron opened the pull request:
Backporting #153 and #157 to Foxy.

In order to maintain backwards compatibility, I've kept the `namespace` argument to `LifecycleNode` as optional (f7c5990).

@ivanpauno PTAL, I *think* this is safe to backport, but you should have more context. I want to backport this, since it is a prerequisite for several other fixes I have made related to launching composable nodes.

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -61,15 +61,18 @@
+        namespace: Optional[SomeSubstitutionsType] = None,
         node_name: Optional[SomeSubstitutionsType] = None,
-        node_namespace: SomeSubstitutionsType = '',
+        node_namespace: SomeSubstitutionsType = None,
```
is this change an API break?

ivanpauno replied:
To me more clear, changing the default isn't the problem.
The problem is that `namespace=''` meant that the namespace in code was going to be used (no remapping rule passed), and now it means that the namespace is remapped to `/` (`__ns:=/` rule passed).

I think that is a (subtle) API break.
I think that the behavior now is the correct one, but I'm not sure if changing the behavior in a previous distro is ok.

jacobperron replied:
Thanks for the review. I've amended the logic so that we don't change the behavior in Foxy (e7e08a7). Now `namespace=None` and `namespace=''` act the same (ie. no remapping rule is passed).

ivanpauno commented on launch_ros/examples/lifecycle_pub_sub_launch.py:
```diff
@@ -36,7 +36,7 @@ def main(argv=sys.argv[1:]):
     # Prepare the talker node.
     talker_node = launch_ros.actions.LifecycleNode(
-        node_name='talker',
+        name='talker', namespace='',
```
The current change is ok, but this other version should still be working:
```suggestion
        name='talker'
```

jacobperron replied:
7ddcc1a

ivanpauno commented on launch_ros/examples/lifecycle_pub_sub_launch.py:
```diff
@@ -62,7 +62,7 @@ def main(argv=sys.argv[1:]):
                     msg="node 'talker' reached the 'active' state, launching 'listener'."),
                 launch_ros.actions.LifecycleNode(
-                    node_name='listener',
+                    name='listener', namespace='',
```
```suggestion
                    name='listener'
```

jacobperron replied:
7ddcc1a

ivanpauno commented on launch_ros/launch_ros/actions/lifecycle_node.py:
```diff
@@ -42,8 +42,9 @@ class LifecycleNode(Node):
-        name: Optional[Text] = None,
-        node_name: Optional[Text] = None,
+        name: Optional[SomeSubstitutionsType] = None,
+        namespace: SomeSubstitutionsType = '',
```
this should be the same:
```suggestion
        namespace: SomeSubstitutionsType = None,
```

jacobperron replied:
7ddcc1a

ivanpauno reviewed (approved):
This seems to fix the bug without breaking API, thanks @jacobperron !

jacobperron commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -348,7 +348,7 @@ def _perform_substitutions(self, context: LaunchContext) -> None:
             self.__expanded_node_name.lstrip('/')
             expanded_node_namespace = None
-            if self.__node_namespace is not None:
+            if self.__node_namespace is not None and self.__node_namespace != '':
```
Done: 7ddcc1a

jacobperron commented:
@ivanpauno I've addressed your comments, and also fixed a bug launching LifecycleNode without a namespace (7ddcc1a). The example was not working, but now it is.

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=12617)](http://ci.ros2.org/job/ci_linux/12617/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=7579)](http://ci.ros2.org/job/ci_linux-aarch64/7579/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=10325)](http://ci.ros2.org/job/ci_osx/10325/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=12543)](http://ci.ros2.org/job/ci_windows/12543/)
